io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -985,38 +1006,62 @@ execAtomically :: ThreadId
                -> TVarId
                -> StmA s a
                -> ST s (StmTxResult s a)
-execAtomically mytid = go [] []
+execAtomically mytid = go [] [] []
   where
-    go :: [SomeTVar s] -> [SomeTVar s] -> TVarId
+    go :: [SomeTVar s]
+       -> [SomeTVar s]
+       -> [(Int, StmA s a)] -- list of checkpoints of written variables at the
+                            -- point of @OrElse@ and second argument of
+                            -- @OrElse@
+       -> TVarId
        -> StmA s a -> ST s (StmTxResult s a)
-    go read written nextVid action = case action of
+    go read written orElses nextVid action = case action of
       ReturnStm x     -> do (vids, tids) <- finaliseCommit written
                             return (StmTxComitted x vids tids nextVid)
       ThrowStm e      -> do finaliseAbort written
                             return (StmTxAborted (toException e))
-      Retry           -> do vids <- finaliseRetry read written
-                            return (StmTxBlocked vids)
+      Retry           -> case orElses of
+                            [] -> do
+                              vids <- finaliseRetry read written
+                              return (StmTxBlocked vids)
+                            (idx, action') : orElses' -> do
+                              -- revert TVar's written since last @OrElse@
+                              finaliseAbortOrElse (drop idx written)
+                              -- and continue evaluating the other branch of
+                              -- @OrElse@
+                              go read written orElses' nextVid action'
       NewTVar x k     -> do v <- execNewTVar nextVid x
-                            go read written (succ nextVid) (k v)
+                            go read written orElses (succ nextVid) (k v)
       ReadTVar v k    -> do x <- execReadTVar v
-                            go (SomeTVar v : read) written nextVid (k x)
-      WriteTVar v x k -> do execWriteTVar v x
-                            go read (SomeTVar v : written) nextVid k
+                            go (SomeTVar v : read) written orElses nextVid (k x)
+      WriteTVar v x k -> do case orElses of
+                              []  -> execWriteTVar v x
+                              _:_ -> execWriteTVarOrElse v x
+                            go read (SomeTVar v : written) orElses nextVid k
+      OrElse x y      -> -- store checkpoints and possible continuation needed
+                         -- in case of a @retry@: to revert the variables
+                         -- modified after this point and jump to the recorded
+                         -- transaction
+                         go read written ((length written, y) : orElses) nextVid x
```

Comment:
I'm having a hard time convincing myself that this saving the length and then later using `drop` is right. I would be more convinced if we had any tests that used more than one variable. Supposing we have this length/index stuff wrong, it's not clear that we would notice since the effects of `execWriteTVarOrElse` and `finaliseAbortOrElse` are idempotent when used on the same variable.

Reply:
> I'm having a hard time convincing myself that ...
That's because it was wrong. We need to track which variables are written in each evaluation branch of `orElse` separately.  I do it using the following type:
```
data St a b where
    NilSt   :: a -> St a b
    LeftSt  :: a -> b -> St a b -> St a b
    RightSt :: a ->      St a b  -> St a b
```
where `a` is instantiated to `[SomeTVar s]` and `b` is stm transaction (right branch of `orElse`, saved for later).

I also made the `prop_nested_orElse` test more robust.  Now the generated trees as structurally folded with either `orElse` or `>>` since there is some interplay between them.  While playing with various parameters I discovered a possible GHC bug for large trees:
```
internal error: Unable to commit 1048576 bytes of memory                    
    (GHC version 8.6.3 for x86_64_unknown_linux)                            
    Please report this as a GHC bug:  http://www.haskell.org/ghc/reportabug 
    17615 abort (core dumped)   
``` 
I think the depth of a tree must be greater than 10 for that, and now we only generate trees of depth <= 8 (since a bug in our implementation would be likely uncovered by a small tree).